python/tk_framework_desktopserver/shotgun/scripts/cache_commands.py

```diff
@@ -0,0 +1,125 @@
+# Copyright (c) 2017 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import sys
+import cPickle
+
+def cache(cache_file, data, base_configuration, hash_data, engine_name):
+    import sqlite3
+    import sgtk
+
+    entity = dict(type=data["entity_type"], id=data["entity_id"])
+    project_entity = dict(id=data["project_id"], type="Project")
+
+    # Setup the bootstrap manager.
+    toolkit_mgr = sgtk.bootstrap.ToolkitManager()
+    toolkit_mgr.plugin_id = "basic.shotgun.cache"
+    toolkit_mgr.base_configuration = base_configuration
+
+    pcs = toolkit_mgr.get_pipeline_configurations(
+        project=project_entity,
+    ) or [dict(id=None)]
+
+    for pc in pcs:
+        lookup_hash = hash_data[pc["id"]]["lookup_hash"]
+        contents_hash = hash_data[pc["id"]]["contents_hash"]
+
+        toolkit_mgr.pipeline_configuration = pc["id"]
+        engine = toolkit_mgr.bootstrap_engine(engine_name, entity=entity)
+        pc_descriptor = toolkit_mgr.resolve_descriptor(project_entity)
+
+        engine.logger.debug("Processing engine commands...")
+        commands = []
+
+        for cmd_name, data in engine.commands.iteritems():
+            engine.logger.debug("Processing command: %s" % cmd_name)
+            props = data["properties"]
+            app = props.get("app")
+
+            if app:
+                app_name = app.name
+            else:
+                app_name = None
+
+            commands.append(
+                dict(
+                    name=cmd_name,
+                    title=props.get("title", cmd_name),
+                    deny_permissions=props.get("deny_permissions", []),
+                    supports_multiple_selection=props.get(
+                        "supports_multiple_selection",
+                        False
+                    ),
+                    app_name=app_name,
+                    group=props.get("group"),
+                    group_default=props.get("group_default"),
+                    engine_name=props.get("engine_name"),
+                ),
+            )
+
+        engine.logger.debug("Engine commands processed.")
+        engine.logger.debug("Inserting commands into cache...")
+
+        # Connect to the database and get the hashes we need to include in
+        # the insert. Each of the lookups call out to the browser_integration
+        # core hook.
+        with sqlite3.connect(cache_file) as connection:
```

Comment:
CalledProcessError is raised when the subprocess fails. This is caught and an error reply is sent to the client, where it's handled and displayed in the action menu.